contrib/checkstyle/src/mill/contrib/checkstyle/CheckstyleModule.scala

```diff
@@ -0,0 +1,183 @@
+    } else if (errs > 0 && os.exists(report)) {
+      val msg = s"checkstyle found $errs error(s), details in $report"
+      if (_throw) {
+        throw new RuntimeException(msg)
+      } else {
+        T.log.error(msg)
+      }
+    } else {
+      throw new RuntimeException(s"checkstyle exit($errs)")
+    }
+
+    val reports = transformer.fold(Seq.empty[PathRef]) { transformer =>
+      val f = transformer(report)
+      transforms.map {
+        case (ref, rel) =>
+          val transform = ref.path
+          val out = dst / rel
+          T.log.info(s"transforming checkstyle report using $transform")
+          f(transform, out)
+          T.log.info(s"transformed checkstyle report to $out")
+          PathRef(out)
+      }.toSeq
+    }
+
+    PathRef(report) +: reports
+  }
+
+  def checkstyleClasspath: T[Loose.Agg[PathRef]] = T {
+    defaultResolver().resolveDeps(
+      Agg(ivy"com.puppycrawl.tools:checkstyle:${checkstyleVersion()}")
+    )
+  }
+
+  def checkstyleConfig: T[PathRef] = T.source {
+    checkstyleDir().path / "config.xml"
+  }
+
+  def checkstyleDir: T[PathRef] = T {
+    PathRef(millSourcePath / "checkstyle")
+  }
+
+  def checkstyleFormat: T[String] = T {
+    "xml"
+  }
+
+  def checkstyleOptions: T[Seq[String]] = T {
+    if (isScala) Seq("-x", ".*[\\.]scala") else Seq.empty
+  }
+
+  def checkstyleReport: T[String] = T {
+    val fmt = checkstyleFormat()
+    val ext = if (fmt == "plain") "txt" else fmt
+    s"report.$ext"
+  }
+
+  /**
+   * A set of [[Transform transformation]]s to be applied on a `checkstyle` report.
+   *
+   * The implementation scans for files under the [[checkstyleDir]] `/` [[checkstyleFormat]].
+   * The selection process is best illustrated with an example.
+   * {{{
+   * /*
+   *
+   * checkstyle
+   *  ├─ xml
+   *  │  ├─ html
+   *  │  │   ├─ xslt0.xml
+   *  │  │   └─ xslt1.xml
+   *  │  └─ pdf
+   *  │      ├─ xslt1.xml
+   *  │      └─ xslt2.xml
+   *  └─ config.xml
+   *
+   * The directory structure above results in the following mapping:
+   *  - checkstyle/xml/html/xslt0.xml -> xslt0.html
+   *  - checkstyle/xml/html/xslt1.xml -> xslt1.html
+   *  - checkstyle/xml/pdf/xslt1.xml  -> xslt1.pdf
+   *  - checkstyle/xml/pdf/xslt2.xml  -> xslt2.pdf
+   *
+   * */
+   * }}}
+   */
+  def checkstyleTransforms: T[Set[Transform]] = T[Set[Transform]] {
+    val dir = checkstyleDir().path / checkstyleFormat()
+
+    if (os.exists(dir)) {
+      T.log.info(s"scanning for transformation files in $dir")
+      os.list(dir)
+        .iterator
+        .filter(os.isDir)
+        .flatMap { ext =>
+          os.list(ext)
+            .iterator
+            .filter(os.isFile)
+            .map(transform => PathRef(transform) -> Seq(s"${transform.baseName}.${ext.baseName}"))
+        }
+        .toSet
+    } else Set.empty[Transform]
+  }
+
+  def checkstyleTransformer(format: String): Option[Transformer] =
+    Option.when(format == "xml")(Transformer.Xml)
```

Comment:
I don't think this needs to be a public method since it's only used inside `checkstyle`

Reply:
Leaving it `public` allows users to plugin their own.
For example, the `sarif` output is a JSON file. A user may chose to run some custom transformation.

Reply:
Rather than having `def checkstyleTransformer(format: String): Option[Transformer]` be an overrideable function that is then used in `def checkstyle` to transform some files, what if we instead break up `def checkstyle` into two separate tasks? 

1. `def checkstyleRaw` that takes generates the initial untransformed files
2. `def checkstyle` that takes the untransformed files and transforms them

In general it's better to have overrideable `T` targets than overrideable helper methods in Mill, because targets can be inspected via the command line or on disk whereas helper methods are opaque

Reply:
This was the original design. I blame my poor naming convention (`checkstyleTransforms` and `checkstyleTransformer`) for the confusion.

The new `getCheckstyleTransformations` function identifies `CheckstyleTransformations` given a `definitionDir` (`checkstyle` configuration directory) and `outputDir` (`checkstyle` task destination directory).

If this is changed to a task, is it possible to access `checkstyle` task destination directory?

Reply:
Actually, the new task can use its own destination directory.  I will make the change.

Reply:
Ok. Yes it should be able to just copy the necessary files into its own directory and return PathRefs to them from there. That way we can skip the entire helper method thing entirely and have the logic live in the target implementations